hydra-node/test/Hydra/Chain/Direct/TxTraceSpec.hs

```diff
@@ -113,80 +263,253 @@ data State
-              pure $ Some $ Close{actor, snapshotNumber}
+      Open{} ->
+        frequency $
+          [
+            ( 1
+            , do
+                actor <- elements allActors
+                snapshot <- genSnapshot
+                pure $ Some $ Close{actor, snapshot}
+            )
           ]
-      Closed ->
-        case maybeGenContest of
-          Nothing -> genFanout
-          Just contestAction -> oneof [contestAction, genFanout]
+            <> [ ( 10
+                 , do
+                    actor <- elements allActors
+                    snapshot <- genSnapshot
+                    pure $ Some Decrement{actor, snapshot}
+                 )
+               | -- XXX: We dont want to generate decrements if there is nothing in the head.
+               not (null utxoInHead)
+               ]
+      Closed{} ->
+        oneof $
+          [ do
+              snapshot <- genSnapshot
+              pure $ Some $ Fanout{snapshot}
+          ]
+            <> [ do
+                  actor <- elements allActors
+                  snapshot <- genSnapshot
+                  pure $ Some Contest{actor, snapshot}
+               ]
       Final -> pure $ Some Stop
    where
-    genFanout = pure $ Some Fanout
-
-    possibleContesters = allActors \\ alreadyContested
-
-    maybeGenContest
-      | null possibleContesters || null snapshots = Nothing
-      | otherwise = Just $ do
-          actor <- elements possibleContesters
-          snapshotNumber <- elements snapshots
-          pure $ Some Contest{actor, snapshotNumber}
-
-  initialState =
-    Model
-      { snapshots = []
-      , headState = Open
-      , utxoV = mkVar (-1)
-      , alreadyContested = []
-      }
+    genSnapshot = do
+      someUTxOToDecrement <- reduceValues =<< genSubModelOf utxoInHead
+      let filteredSomeUTxOToDecrement = Map.filter (> 0) someUTxOToDecrement
+      let balancedUTxOInHead = balanceUTxOInHead utxoInHead filteredSomeUTxOToDecrement
+      let validSnapshot =
+            ModelSnapshot
+              { snapshotNumber = latestSnapshot
+              , snapshotUTxO = balancedUTxOInHead
+              , decommitUTxO = filteredSomeUTxOToDecrement
+              }
+      oneof
+        [ -- valid
+          pure validSnapshot
+        , -- unbalanced
+          pure validSnapshot{snapshotUTxO = utxoInHead}
+        , do
+            -- old
+            let snapshotNumber' = if latestSnapshot == 0 then 0 else latestSnapshot - 1
+            pure validSnapshot{snapshotNumber = snapshotNumber'}
+        , -- new
+          pure validSnapshot{snapshotNumber = latestSnapshot + 1}
+        , do
+            -- shuffled
+            someUTxOToDecrement' <- shuffleValues filteredSomeUTxOToDecrement
+            pure validSnapshot{decommitUTxO = someUTxOToDecrement'}
+        , do
+            -- more in head
+            utxoInHead' <- increaseValues utxoInHead
+            pure validSnapshot{snapshotUTxO = utxoInHead'}
```

Comment:
Aren't these invalid? Do we really want to create invalid transitions? IIRC this test suite specializes on testing sequences of otherwise valid steps